text/0171-cloudfront-redesign.md

```diff
@@ -0,0 +1,489 @@
+    edgeFunctions: [{
+      functionVersion: myFunc.currentVersion,
+      eventType: EventType.VIEWER_REQUEST,
+    }],
+  }),
+});
+```
+
+Lambda@Edge functions can also be associated with additional behaviors, either at behavior creation (associated with the origin) or after behavior
+creation.
+
+```ts
+// Assigning at behavior creation.
+myOrigin.addBehavior('/images/*.jpg', {
+  viewerProtocolPolicy: ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
+  defaultTtl: cdk.Duration.days(7),
+  edgeFunctions: [{
+    functionVersion: myFunc.currentVersion,
+    eventType: EventType.VIEWER_REQUEST,
+  }]
+});
+
+// Assigning after creation.
+const myImagesBehavior = myOrigin.addBehavior('/images/*.jpg', ...);
+myImagesBehavior.addEdgeFunction({
+  functionVersion: myFunc.currentVersion,
+  eventType: EventType.VIEWER_REQUEST,
+});
+```
+
+---
+
+# Motivation
+
+The existing aws-cloudfront module doesn't adhere to standard naming convention, lacks convenience methods for more easily interacting with
+distributions, origins, and behaviors, and has been in an "experimental" state for years. This proposal aims to bring a friendlier, more ergonomic
+interface to the module, and advance the module to a GA-ready state.
+
+# Design Summary
+
+The approach will create a new top-level Construct (`Distribution`) to replace the existing `CloudFrontWebDistribution`, as well as new constructs to
+represent the other logical resources for a distribution (i.e., `Origin`, `Behavior`). The new construct is optimized for the most common use cases of
+creating a distribution with a single origin and behavior. The new L2s will be created in the same aws-cloudfront module and no changes will be made
+to the existing L2s to preserve the existing experience. Unlike the existing L2, the new L2s will feature a variety of convenience methods (e.g.,
+`addBehavior`) to aid in the creation of the distribution, and provide several out-of-the-box defaults for building distributions off of other
+resources (e.g., buckets, load balanced services).
+
+# Detailed Design
+
+The design creates one new resource (`Distribution`) and two new classes (`Origin` and `Behavior`) to replace the current
+`CloudFrontWebDistribution` construct. Each of these new classes comes with helper methods (e.g., `addBehavior`) to make assembling more complex
+distributions easier, as well as factory constructors to make it easier to build common Origin and Behavior patterns (e.g., `Origin.fromBucket`).
+
+The following is an incomplete, but representative, listing of the API:
+
+```ts
+class Distribution extends BaseDistribution {
+  static fromArn(scope: Construct, id: string, distributionArn: string): IDistribution;
+  static fromAttributes(scope: Construct, id: string, distributionArn: string): IDistribution;
+
+  static forBucket(scope: Construct, id: string, bucket: IBucket): IDistribution;
+  static forWebsiteBucket(scope: Construct, id: string, bucket: IBucket): IDistribution;
```

Comment:
Isn't this a problem? If `forBucket` and `forWebiteBucket` returns an `IDistribution`, that interface doesn't have the `addOrigin` method, so you can't add additional origins to a Distribution created that way.

Reply:
Yup, bad copy/paste. Those should definitely both return a Distribution, not IDistribution.